File: src/4node/classify/nodes/classifyNode.ts
// ─── classifyNode — 意图分类 ───────────────────────────────────────────────
// 归属：主图节点

import { HumanMessage } from '@langchain/core/messages'
import { z } from 'zod'
import { createChatModel } from '../../../2model/model.js'
import { classifyPrompt } from '../../../1common/prompts.js'
import type { MainState } from '../../../3state/index.js'

const ClassifySchema = z.object({
  intent: z.enum(['consult', 'apply', 'insufficient'])
          .describe('如果是咨询政策为consult；如果想要申请加分但欠缺赛事名称/时间/等级信息为insufficient；如果要申请且信息完整为apply'),
  missing: z.array(z.string())
            .describe('只有在 intent 为 insufficient 时，列出缺失的字段'),
  documentText: z.string()
                .describe('只有在 intent 为 apply 时，提取用户用来申请的完整材料原文'),
})

export async function classifyNode(state: MainState): Promise<Partial<MainState>> {
  // 申请入口优先：forcedIntent 直接指定 intent，跳过 LLM 分类
  if (state.forcedIntent) {
    console.log(`-main:classifyNode: forcedIntent=${state.forcedIntent}，跳过 LLM 分类`)
    return { intent: state.forcedIntent }
  }

  const allUserText = state.messages
    .filter(m => m instanceof HumanMessage)
    .map(m => String(m.content))
    .join('\n')

  const model = createChatModel(0).withStructuredOutput(ClassifySchema)
  const reply = await model.invoke([new HumanMessage(classifyPrompt(allUserText))])

  console.log(`-main:classifyNode: 意图=${reply.intent}, 缺失=${reply.missing}, 材料=${reply.documentText?.slice(0, 10)}...`)

  return {
    intent: reply.intent,
    missingInfo: reply.missing || [],
    documentText: reply.documentText || '',
  }
}

File: src/nodes/classify/classify_node.py
"""classifyNode - 意图分类 - 对应 TS: classifyNode.ts"""
import json
from typing import Literal
from pydantic import BaseModel
from langchain_core.messages import HumanMessage
from langchain_core.messages import BaseMessage
from models.llm import create_chat_model
from prompts.templates import classify_prompt


class ClassifyOutput(BaseModel):
    """意图分类输出"""
    intent: Literal["consult", "apply", "insufficient"]
    missing: list[str] = []
    document_text: str = ""


async def classify_node(state: dict) -> dict:
    """
    意图分类节点
    
    归属: 主图节点
    功能: 分析用户输入，判断意图并提取信息
    
    forcedIntent 优先：申请入口专用，优先级高于 LLM 分类
    """
    # forcedIntent 优先
    forced_intent = state.get("forced_intent")
    if forced_intent:
        print(f"-main:classifyNode: forcedIntent={forced_intent}，跳过 LLM 分类")
        return {"intent": forced_intent}
    
    # 提取用户消息
    messages: list[BaseMessage] = state.get("messages", [])
    all_user_text = "\n".join([
        m.content for m in messages
        if isinstance(m, HumanMessage)
    ])
    
    if not all_user_text:
        return {"intent": "consult"}
    
    # LLM 分类
    model = create_chat_model(temperature=0)
    structured_model = model.with_structured_output(ClassifyOutput)
    
    result = await structured_model.ainvoke([
        HumanMessage(content=classify_prompt(all_user_text))
    ])
    
    print(f"-main:classifyNode: 意图={result.intent}, 缺失={result.missing}, 材料={result.document_text[:10] if result.document_text else ''}...")
    
    return {
        "intent": result.intent,
        "missing_info": result.missing or [],
        "document_text": result.document_text or "",
    }
